### DISCUSSION TITLE
Why is no value being produced when I transform an observable into a flow and launch it in viewModelScope within the init block?

### DISCUSSION THREAD

**Discussion by hvsimon** [cashapp/turbine#232] [Q&A] [answered]
In the ViewModel, I would like to load data in the init block and update the state flow so that the UI can display it.

However, our model layer returns observables and has not been converted to coroutines yet due to certain reasons. 
But in the ViewModel layer, I want to use coroutines. Therefore, I handle it by converting the observables to flows using `asFlow`.

The problem is that when I test whether the state flow has the correct data, I noticed that the initial value and the data in onStart are correct, but the data modified in onEach does not appear.

Here is an example code:

```
@HiltViewModel
class MyViewModel @Inject constructor(
    myRepository: MyRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow(0)
    val uiState = _uiState.asStateFlow()

    init {
        myRepository.getData() // return io.reactivex.rxjava3.core.Observable
            .asFlow()
            .onStart { _uiState.update { 1 } }
            .onEach { data ->
                _uiState.update { data }
            }
            .launchIn(viewModelScope)
    }
}
```

```
@OptIn(ExperimentalCoroutinesApi::class)
class MyViewModelTest {

    private lateinit var viewModel: MyViewModel
    private val myRepository = mockk<MyRepository>(relaxed = true)

    @Before
    fun setup() {
        Dispatchers.setMain(StandardTestDispatcher())
        viewModel = MyViewModel(myRepository)
    }

    @After
    fun teardown() {
        Dispatchers.resetMain()
    }

    @Test
    fun fooBar() = runTest {

        every { myRepository.getData() } returns Observable.just(2)

        viewModel.uiState.test {
            // init
            assertEquals(0, awaitItem()) // Passed

            // onStart
            assertEquals(1, awaitItem()) // Passed

            // onEach
            assertEquals(2, awaitItem()) // No value produced in 1s
        }

        advanceUntilIdle()
        verify { myRepository.getData() }
    }
}
```

**Comment by jingibus** [comment]
I'm unsure whether this points to a specific issue in Turbine, or to an issue in your testing setup, which is not something we provide support for as Turbine maintainers.

If you can provide a smaller repro that points to a particular bug in Turbine, please share it as an issue. Thank you!

**Reply by hvsimon** [reply]
This might not be an issue report, I'm actually wondering why the last `assertEquals` cannot pass. The above code can reproduce it.

**Reply by jingibus** [reply, marked as answer]
Well, you aren't using Turbine wrong or anything. You are using Turbine correctly, but your test setup is failing to plug up `Observable.just(2)`. By the time this line has run,

```
every { myRepository.getData() } returns Observable.just(2)
```

...`myRepository.getData()` has already run. 

So: not really a question about Turbine, but more a question about "How do I do test setup with coroutines?"

Launching work in `init` is pretty common. If the subject is constructed inside a `@Before`, however, then it's not possible to configure any dependencies used by that work: by the time you get to `@Test`, the dependencies have already been invoked. So I prefer to avoid using `@Before` methods to construct the subject under test.
